sdk/storage/azure-storage-blob/azure/storage/blob/download.py

```diff
@@ -192,227 +207,203 @@ def _download_chunk(self, chunk_start, chunk_end):
             )
 
-        if content is not None:
-            yield content
-        if self._download_complete:
-            return
-
-        data_end = self.file_size
-        if self.length is not None:
-            # Use the length unless it is over the end of the file
-            data_end = min(self.file_size, self.length + 1)
-
-        downloader = SequentialChunkDownloader(
-            client=self.clients.blob,
-            non_empty_ranges=self.non_empty_ranges,
-            total_size=self.download_size,
-            chunk_size=self.config.max_chunk_get_size,
-            current_progress=self.first_get_size,
-            start_range=self.initial_range[1] + 1,  # start where the first download ended
-            end_range=data_end,
-            stream=None,
-            validate_content=self.validate_content,
-            encryption_options=self.encryption_options,
-            use_location=self.location_mode,
-            **self.request_options
-        )
-
-        for chunk in downloader.get_chunk_offsets():
-            yield downloader.yield_chunk(chunk)
+    def __len__(self):
+        return self.size
 
     def _initial_request(self):
         range_header, range_validation = validate_and_format_range_headers(
-            self.initial_range[0],
-            self.initial_range[1],
+            self._initial_range[0],
+            self._initial_range[1],
             start_range_required=False,
             end_range_required=False,
-            check_content_md5=self.validate_content
+            check_content_md5=self._validate_content
         )
 
         try:
-            location_mode, response = self.clients.blob.download(
+            location_mode, response = self._clients.blob.download(
                 range=range_header,
                 range_get_content_md5=range_validation,
-                validate_content=self.validate_content,
+                validate_content=self._validate_content,
                 data_stream_total=None,
                 download_stream_current=0,
-                **self.request_options
+                **self._request_options
             )
 
             # Check the location we read from to ensure we use the same one
             # for subsequent requests.
-            self.location_mode = location_mode
+            self._location_mode = location_mode
 
             # Parse the total file size and adjust the download size if ranges
             # were specified
-            self.file_size = parse_length_from_content_range(response.properties.content_range)
-            if self.length is not None:
-                # Use the length unless it is over the end of the file
-                self.download_size = min(self.file_size, self.length - self.offset + 1)
-            elif self.offset is not None:
-                self.download_size = self.file_size - self.offset
+            self._file_size = parse_length_from_content_range(response.properties.content_range)
```

Comment:
I have a question about file_size. If we uploaded a 30 bytes file with client side encryption, then the encrypted file is 32 bytes. When we want to download the file, the parsed self._file_size is 32 I assume, are we intended to keep 32?

Reply:
The file size here is purely with regards to the bytes being downloaded, and the values used in the range header. So if the encrypted file is 32, that will be the value we use. If this does not match the decrypted value size, we don't update it... I think that's valid, as the size attribute is to determine the size of the Stream (i.e. what's being downloaded).